File: cli/node/src/output/format.ts
import type { JsonObject } from '../backend/json.js';
import { parseJsonValue } from '../backend/json.js';
/** CLI 文本、JSON、表格和静默输出。 */

import boxen from 'boxen';
import Table from 'cli-table3';
import { takeNotice } from '../runtime/state.js';
import { colors, sym } from './branding.js';

const { brand, accent, success, error: errorColor, dim } = colors;

/**
 * 提取日期部分；无法解析时保留原字符串的前十位。
 * @param dtStr - 待展示的日期字符串。
 * @returns 日期文本；空输入返回未定义。
 */
function formatDate(dtStr?: string): string | undefined {
  if (!dtStr) return undefined;
  try {
    const dt = new Date(dtStr.replace('Z', '+00:00'));
    return dt.toISOString().slice(0, 10);
  } catch {
    return dtStr?.slice(0, 10);
  }
}

/**
 * 将记忆列表输出为带摘要的终端文本。
 * @param memories - 后端返回的记忆列表。
 * @param title - 展示标题。
 */
export function formatMemoriesText(memories: JsonObject[], title: string): void {
  const count = memories.length;
  console.log(`\n${brand(`Found ${count} ${title}:`)}\n`);

  for (let i = 0; i < memories.length; i++) {
    const mem = memories[i];
    const memoryText = (mem.memory ?? mem.text ?? '') as string;
    const memId = ((mem.id as string) ?? '').slice(0, 8);
    const score = mem.score as number | undefined;
    const created = formatDate(mem.created_at as string | undefined);
    let category: string | undefined;
    const cats = mem.categories;
    if (Array.isArray(cats)) {
      category = cats[0] as string | undefined;
    }

    console.log(`  ${i + 1}. ${memoryText}`);

    const details: string[] = [];
    if (score !== undefined) details.push(`Score: ${score.toFixed(2)}`);
    if (memId) details.push(`ID: ${memId}`);
    if (created) details.push(`Created: ${created}`);
    if (category) details.push(`Category: ${category}`);

    if (details.length > 0) {
      console.log(`     ${dim(details.join(' · '))}`);
    }
    console.log();
  }
}

/**
 * 将记忆列表输出为表格。
 * @param memories - 后端返回的记忆列表。
 * @param opts - 本次操作选项。
 * @param opts.showScore - 在表格中显示匹配分数。
 */
export function formatMemoriesTable(memories: JsonObject[], opts: { showScore?: boolean }): void {
  const head = opts.showScore
    ? [accent('ID'), accent('Score'), accent('Memory'), accent('Category'), accent('Created')]
    : [accent('ID'), accent('Memory'), accent('Category'), accent('Created')];
  const colWidths = opts.showScore ? [38, 8, 40, 16, 14] : [38, 40, 16, 14];
  const table = new Table({
    head,
    colWidths,
    wordWrap: true,
    style: { head: [], border: [] },
  });

  for (const mem of memories) {
    const memId = (mem.id as string) ?? '';
    let memoryText = (mem.memory ?? mem.text ?? '') as string;
    if (memoryText.length > 60) {
      memoryText = `${memoryText.slice(0, 57)}...`;
    }
    const categories = mem.categories;
    const cat =
      Array.isArray(categories) && categories.length > 0
        ? categories.length > 1
          ? `${categories[0]} (+${categories.length - 1})`
          : (categories[0] as string)
        : '—';
    const created = formatDate(mem.created_at as string | undefined) ?? '—';
    if (opts.showScore) {
      const score = mem.score as number | undefined;
      const scoreStr = score !== undefined ? score.toFixed(2) : '—';
      table.push([dim(memId), scoreStr, memoryText, cat, created]);
    } else {
      table.push([dim(memId), memoryText, cat, created]);
    }
  }

  console.log();
  console.log(table.toString());
  console.log();
}

/**
 * 将数据序列化为缩进 JSON 并输出。
 * @param data - 需要输出的数据。
 */
export function formatJson(data: unknown): void {
  console.log(JSON.stringify(data, null, 2));
}

/**
 * 按所选格式输出单条记忆详情。
 * @param mem - 单条记忆对象。
 * @param output - 终端输出格式。
 */
export function formatSingleMemory(mem: JsonObject, output: string): void {
  if (output === 'json') {
    formatJson(mem);
    return;
  }

  const memoryText = (mem.memory ?? mem.text ?? '') as string;
  const memId = (mem.id ?? '') as string;

  const lines: string[] = [];
  lines.push(`  ${memoryText}`);
  lines.push('');

  if (memId) lines.push(`  ${dim('ID:')}         ${memId}`);
  const created = formatDate(mem.created_at as string | undefined);
  if (created) lines.push(`  ${dim('Created:')}    ${created}`);
  const updated = formatDate(mem.updated_at as string | undefined);
  if (updated) lines.push(`  ${dim('Updated:')}    ${updated}`);
  const meta = mem.metadata;
  if (meta) lines.push(`  ${dim('Metadata:')}   ${JSON.stringify(meta)}`);
  const categories = mem.categories;
  if (categories) {
    const catStr = Array.isArray(categories) ? categories.join(', ') : String(categories);
    lines.push(`  ${dim('Categories:')} ${catStr}`);
  }

  const content = lines.join('\n');
  console.log();
  console.log(
    boxen(content, {
      title: brand('Memory'),
      titleAlignment: 'left',
      borderColor: 'magenta',
      padding: 1,
    })
  );
  console.log();
}

/**
 * 展示添加结果，合并相同事件的待处理提示。
 * @param result - 添加记忆的响应对象或列表。
 * @param output - 终端输出格式。
 */
export function formatAddResult(result: JsonObject | JsonObject[], output: string): void {
  if (output === 'json') {
    formatJson(result);
    return;
  }
  if (output === 'quiet') return;

  const results: JsonObject[] = Array.isArray(result)
    ? result
    : ((result.results as JsonObject[]) ?? [result]);

  if (!results.length) {
    console.log(`  ${dim('No memories extracted.')}`);
    return;
  }

  console.log();
  const seenPendingEvents = new Set<string>();
  for (const r of results) {
    // 识别异步 PENDING 响应。
    if (r.status === 'PENDING') {
      const eventId = (r.event_id as string) ?? '';
      // 相同 event_id 的待处理项只显示一次。
      if (eventId && seenPendingEvents.has(eventId)) continue;
      if (eventId) seenPendingEvents.add(eventId);
      const icon = accent(sym('⧗', '...'));
      const parts = [`  ${icon} ${dim('Queued'.padEnd(10))}`, 'Processing in background'];
      console.log(parts.join('  '));
      if (eventId) {
        console.log(`  ${dim(`  event_id: ${eventId}`)}`);
        console.log(`  ${dim(`  → Check status: memgo event status ${eventId}`)}`);
      }
      continue;
    }

    const event = (r.event ?? 'ADD') as string;
    const memory = (r.memory ?? r.text ?? r.content ?? r.data ?? '') as string;
    const memId = ((r.id as string) ?? (r.memory_id as string) ?? '').slice(0, 8);

    let icon: string;
    let label: string;
    if (event === 'ADD') {
      icon = success('+');
      label = 'Added';
    } else if (event === 'UPDATE') {
      icon = accent('~');
      label = 'Updated';
    } else if (event === 'DELETE') {
      icon = errorColor('-');
      label = 'Deleted';
    } else if (event === 'NOOP') {
      icon = dim('·');
      label = 'No change';
    } else {
      icon = dim('?');
      label = event;
    }

    const parts = [`  ${icon} ${dim(label.padEnd(10))}`];
    if (memory) parts.push(memory);
    if (memId) parts.push(dim(`(${memId})`));
    console.log(parts.join('  '));
  }
  console.log();
}

/**
 * 构造命令 JSON 信封并附加待展示的平台通知。
 * @param opts - 本次操作选项。
 * @param opts.command - 当前命令名称。
 * @param opts.data - 需要输出的数据。
 * @param opts.durationMs - 操作耗时，单位为毫秒。
 * @param opts.scope - 本次操作的实体范围。
 * @param opts.count - 结果数量。
 * @param opts.status - 业务执行状态。
 * @param opts.error - 需要输出的错误描述。
 */
export function formatJsonEnvelope(opts: {
  command: string;
  data: unknown;
  durationMs?: number;
  scope?: Record<string, string | undefined>;
  count?: number;
  status?: string;
  error?: string;
}): void {
  const envelope: JsonObject = {
    status: opts.status ?? 'success',
    command: opts.command,
  };
  if (opts.durationMs !== undefined) envelope.duration_ms = opts.durationMs;
  if (opts.scope !== undefined) envelope.scope = opts.scope;
  if (opts.count !== undefined) envelope.count = opts.count;
  if (opts.error) envelope.error = opts.error;
  envelope.data = opts.data === undefined ? undefined : parseJsonValue(opts.data);

  // 未认领账号提示放入 JSON 信封，调用方无需读取 HTTP 头。

  const notice = takeNotice();
  if (notice) envelope.memgo_notice = notice;

  console.log(JSON.stringify(envelope, null, 2));
}

/**
 * 复制指定字段，忽略不存在的键。
 * @param obj - 字段来源对象。
 * @param keys - 需要保留的字段名。
 * @returns 只包含指定已存在字段的新对象。
 */
function pick(obj: JsonObject, keys: string[]): JsonObject {
  const result: JsonObject = {};
  for (const key of keys) {
    if (key in obj) result[key] = obj[key];
  }
  return result;
}

/**
 * 按命令保留代理输出需要的字段。
 * @param command - 当前命令名称。
 * @param data - 需要输出的数据。
 * @returns 按命令裁剪字段后的数据。
 */
export function sanitizeAgentData(command: string, data: unknown): unknown {
  if (data === null || data === undefined) return data;

  switch (command) {
    case 'add': {
      const items = Array.isArray(data) ? data : [data];
      return items.map((item) => {
        const r = item as JsonObject;
        if (r.status === 'PENDING') return pick(r, ['status', 'event_id']);
        return pick(r, ['id', 'memory', 'event']);
      });
    }
    case 'search':
      return (data as JsonObject[]).map((r) =>
        pick(r, ['id', 'memory', 'score', 'created_at', 'categories', 'expiration_date'])
      );
    case 'list':
      return (data as JsonObject[]).map((r) =>
        pick(r, ['id', 'memory', 'created_at', 'categories', 'expiration_date'])
      );
    case 'get': {
      const r = data as JsonObject;
      return pick(r, [
        'id',
        'memory',
        'created_at',
        'updated_at',
        'categories',
        'metadata',
        'expiration_date',
      ]);
    }
    case 'update': {
      const r = data as JsonObject;
      return pick(r, ['id', 'memory', 'expiration_date']);
    }
    case 'delete':
    case 'delete-all':
    case 'entity delete':
      return data;
    case 'entity list':
      return (data as JsonObject[]).map((r) => ({
        name: (r.name ?? r.id) as string,
        ...pick(r, ['type', 'count']),
      }));
    case 'event list':
      return (data as JsonObject[]).map((r) =>
        pick(r, ['id', 'event_type', 'status', 'latency', 'created_at'])
      );
    case 'event status': {
      const ev = data as JsonObject;
      const rawResults = (ev.results as JsonObject[] | undefined) ?? [];
      const sanitizedResults = rawResults.map((r) => {
        const nested = r.data as JsonObject | undefined;
        return {
          id: r.id,
          event: r.event,
          user_id: r.user_id,
          memory: nested?.memory ?? null,
        };
      });
      return {
        ...pick(ev, ['id', 'event_type', 'status', 'latency', 'created_at', 'updated_at']),
        results: sanitizedResults,
      };
    }
    default:
      return data;
  }
}

/**
 * 输出精简的代理 JSON 信封。
 * @param opts - 本次操作选项。
 * @param opts.command - 当前命令名称。
 * @param opts.data - 需要输出的数据。
 * @param opts.durationMs - 操作耗时，单位为毫秒。
 * @param opts.scope - 本次操作的实体范围。
 * @param opts.count - 结果数量。
 */
export function formatAgentEnvelope(opts: {
  command: string;
  data: unknown;
  durationMs?: number;
  scope?: Record<string, string | undefined>;
  count?: number;
}): void {
  const envelope: JsonObject = {
    status: 'success',
    command: opts.command,
  };
  if (opts.durationMs !== undefined) envelope.duration_ms = opts.durationMs;
  if (opts.scope) {
    const filtered = Object.fromEntries(Object.entries(opts.scope).filter(([, v]) => v));
    if (Object.keys(filtered).length > 0) envelope.scope = filtered;
  }
  if (opts.count !== undefined) envelope.count = opts.count;
  envelope.data = parseJsonValue(sanitizeAgentData(opts.command, opts.data));

  // 在 JSON 信封中显示未认领账号提示。

  const notice = takeNotice();
  if (notice) envelope.memgo_notice = notice;

  console.log(JSON.stringify(envelope, null, 2));
}

/**
 * 展示结果数量、分页、实体范围及耗时摘要。
 * @param opts - 本次操作选项。
 * @param opts.count - 结果数量。
 * @param opts.durationSecs - 操作耗时，单位为秒。
 * @param opts.page - 分页页码。
 * @param opts.scopeIds - 用于摘要的实体范围。
 */
export function printResultSummary(opts: {
  count: number;
  durationSecs?: number;
  page?: number;
  scopeIds?: Record<string, string | undefined>;
}): void {
  const parts = [`${opts.count} result${opts.count !== 1 ? 's' : ''}`];
  if (opts.page !== undefined) parts.push(`page ${opts.page}`);
  if (opts.scopeIds) {
    const scopeParts = Object.entries(opts.scopeIds)
      .filter(([, v]) => v)
      .map(([k, v]) => `${k}=${v}`);
    if (scopeParts.length > 0) parts.push(scopeParts.join(', '));
  }
  if (opts.durationSecs !== undefined) parts.push(`${opts.durationSecs.toFixed(2)}s`);

  console.log(`  ${dim(parts.join(' · '))}`);
  console.log();
}

File: cli/python/src/memgo_cli/output/format.py
"""文本、表格、JSON 和静默输出格式。"""

from __future__ import annotations

import json
from collections.abc import Mapping
from datetime import datetime

from rich.console import Console
from rich.panel import Panel
from rich.table import Table
from rich.text import Text

from memgo_cli.backend.json import (
    JsonObject,
    JsonValue,
    json_object,
    json_records,
    json_string,
    json_value,
)
from memgo_cli.output.branding import ACCENT_COLOR, BRAND_COLOR, DIM_COLOR, SUCCESS_COLOR, _sym


def format_memories_text(console: Console, memories: list[JsonObject], title: str) -> None:
    """以易读文本显示记忆列表。"""
    count = len(memories)
    console.print(f"\n[{BRAND_COLOR}]Found {count} {title}:[/]\n")

    for i, mem in enumerate(memories, 1):
        memory_text = json_string(mem.get("memory") or mem.get("text") or "", "memory.text")
        mem_id = json_string(mem.get("id") or "", "memory.id")[:8]
        score = mem.get("score")
        created = _format_date(mem.get("created_at"))
        category = mem.get("categories", [None])
        if isinstance(category, list):
            category = category[0] if category else None

        line = Text()
        line.append(f"  {i}. ", style="bold")
        line.append(memory_text, style="white")
        console.print(line)

        details = []
        if score is not None:
            details.append(f"Score: {score:.2f}")
        if mem_id:
            details.append(f"ID: {mem_id}")
        if created:
            details.append(f"Created: {created}")
        if category:
            details.append(f"Category: {category}")

        if details:
            detail_str = " · ".join(details)
            console.print(f"     [{DIM_COLOR}]{detail_str}[/]")
        console.print()


def format_memories_table(
    console: Console, memories: list[JsonObject], *, show_score: bool
) -> None:
    """以 Rich 表格显示记忆列表。"""
    table = Table(
        border_style=BRAND_COLOR,
        header_style=f"bold {ACCENT_COLOR}",
        row_styles=["", "dim"],
        padding=(0, 1),
    )
    table.add_column("ID", style="dim", max_width=38, no_wrap=True)
    if show_score:
        table.add_column("Score", max_width=7, justify="right")
    table.add_column("Memory", max_width=50, no_wrap=False)
    table.add_column("Category", max_width=14)
    table.add_column("Created", max_width=12)

    for mem in memories:
        mem_id = json_string(mem.get("id") or "", "memory.id")
        memory_text = json_string(mem.get("memory") or mem.get("text") or "", "memory.text")
        if len(memory_text) > 60:
            memory_text = memory_text[:57] + "..."
        categories = mem.get("categories", [])
        if isinstance(categories, list) and categories:
            cat = (
                json_string(categories[0], "memory.categories")
                if len(categories) == 1
                else f"{categories[0]} (+{len(categories) - 1})"
            )
        else:
            cat = "—"
        created = _format_date(mem.get("created_at")) or "—"
        if show_score:
            score = mem.get("score")
            score_str = f"{score:.2f}" if score is not None else "—"
            table.add_row(mem_id, score_str, memory_text, cat, created)
        else:
            table.add_row(mem_id, memory_text, cat, created)

    console.print()
    console.print(table)
    console.print()


def format_json(console: Console, data: object) -> None:
    """输出带缩进的 JSON 数据。"""
    console.print_json(json.dumps(data, default=str))


def format_single_memory(console: Console, mem: JsonObject, output: str) -> None:
    """显示单条记忆及其元数据。"""
    if output == "json":
        format_json(console, mem)
        return

    memory_text = json_string(mem.get("memory") or mem.get("text") or "", "memory.text")
    mem_id = json_string(mem.get("id") or "", "memory.id")

    lines = []
    lines.append(f"  [white bold]{memory_text}[/]")
    lines.append("")

    if mem_id:
        lines.append(f"  [{DIM_COLOR}]ID:[/]         {mem_id}")
    created = _format_date(mem.get("created_at"))
    if created:
        lines.append(f"  [{DIM_COLOR}]Created:[/]    {created}")
    updated = _format_date(mem.get("updated_at"))
    if updated:
        lines.append(f"  [{DIM_COLOR}]Updated:[/]    {updated}")
    meta = mem.get("metadata")
    if meta:
        lines.append(f"  [{DIM_COLOR}]Metadata:[/]   {json.dumps(meta)}")
    categories = mem.get("categories")
    if categories:
        cat_str = (
            ", ".join(json_string(item, "memory.categories") for item in categories)
            if isinstance(categories, list)
            else categories
        )
        lines.append(f"  [{DIM_COLOR}]Categories:[/] {cat_str}")

    content = "\n".join(lines)
    panel = Panel(
        content,
        title=f"[{BRAND_COLOR}]Memory[/]",
        title_align="left",
        border_style=BRAND_COLOR,
        padding=(1, 1),
    )
    console.print()
    console.print(panel)
    console.print()


def format_add_result(console: Console, result: JsonObject | list[JsonObject], output: str) -> None:
    """显示添加结果，合并同一事件的待处理项。"""
    if output == "json":
        format_json(console, result)
        return
    if output == "quiet":
        return

    # API 通常返回包含 results 数组的对象
    results = json_records(result if isinstance(result, list) else result.get("results", [result]))
    if not results:
        console.print(f"  [{DIM_COLOR}]No memories extracted.[/]")
        return

    console.print()
    seen_pending_events: set[str] = set()
    for r in results:
        # 识别平台异步 PENDING 响应
        if r.get("status") == "PENDING":
            event_id = json_string(r.get("event_id", ""), "event_id")
            # 按 event_id 合并待处理项
            if event_id and event_id in seen_pending_events:
                continue
            if event_id:
                seen_pending_events.add(event_id)
            icon = f"[{ACCENT_COLOR}]{_sym('⧗', '...')}[/]"
            parts = [f"  {icon} [{DIM_COLOR}]{'Queued':<10}[/]"]
            parts.append("[white]Processing in background[/]")
            console.print("  ".join(parts))
            if event_id:
                console.print(f"  [{DIM_COLOR}]  event_id: {event_id}[/]")
                console.print(f"  [{DIM_COLOR}]  → Check status: memgo event status {event_id}[/]")
            continue

        event = json_string(r.get("event", "ADD"), "event")
        memory = r.get("memory") or r.get("text") or r.get("content") or r.get("data") or ""
        mem_id = json_string(r.get("id") or r.get("memory_id") or "", "memory.id")[:8]

        if event == "ADD":
            icon = f"[{SUCCESS_COLOR}]+[/]"
            label = "Added"
        elif event == "UPDATE":
            icon = f"[{ACCENT_COLOR}]~[/]"
            label = "Updated"
        elif event == "DELETE":
            icon = "[red]-[/]"
            label = "Deleted"
        elif event == "NOOP":
            icon = f"[{DIM_COLOR}]·[/]"
            label = "No change"
        else:
            icon = f"[{DIM_COLOR}]?[/]"
            label = event

        # 构造结果显示行
        parts = [f"  {icon} [{DIM_COLOR}]{label:<10}[/]"]
        if memory:
            parts.append(f"[white]{memory}[/]")
        if mem_id:
            parts.append(f"[{DIM_COLOR}]({mem_id})[/]")
        console.print("  ".join(parts))
    console.print()


def format_json_envelope(
    console: Console,
    *,
    command: str,
    data: object,
    duration_ms: int | None,
    scope: Mapping[str, JsonValue] | None,
    count: int | None,
    status: str,
    error: str | None,
) -> None:
    """输出包含状态、数据和平台通知的 JSON 信封。"""
    envelope: JsonObject = {
        "status": status,
        "command": command,
    }
    if duration_ms is not None:
        envelope["duration_ms"] = duration_ms
    if scope is not None:
        envelope["scope"] = scope
    if count is not None:
        envelope["count"] = count
    if error:
        envelope["error"] = error
    envelope["data"] = json_value(data)

    # 平台返回未认领账号提示时，将其加入 JSON 信封
    # 代理读取输出即可获得提示
    # 无需检查 HTTP 响应头
    from memgo_cli.runtime.state import take_notice

    notice = take_notice()
    if notice:
        envelope["memgo_notice"] = notice

    console.print_json(json.dumps(envelope, default=str))


def sanitize_agent_data(command: str, data: object) -> object:
    """按命令提取代理需要的响应字段。"""

    def pick(obj: JsonObject, keys: list[str]) -> JsonObject:
        """提取指定字段，忽略响应中的其他数据。"""
        return {k: obj[k] for k in keys if k in obj}

    if data is None:
        return data

    if command == "add":
        items = json_records(data if isinstance(data, list) else [data])
        result = []
        for item in items:
            if item.get("status") == "PENDING":
                result.append(pick(item, ["status", "event_id"]))
            else:
                result.append(pick(item, ["id", "memory", "event"]))
        return result

    if command == "search":
        return [
            pick(r, ["id", "memory", "score", "created_at", "categories", "expiration_date"])
            for r in json_records(data)
        ]

    if command == "list":
        return [
            pick(r, ["id", "memory", "created_at", "categories", "expiration_date"])
            for r in json_records(data)
        ]

    if command == "get":
        return pick(
            json_object(data),
            [
                "id",
                "memory",
                "created_at",
                "updated_at",
                "categories",
                "metadata",
                "expiration_date",
            ],
        )

    if command == "update":
        return pick(json_object(data), ["id", "memory", "expiration_date"])

    if command in ("delete", "delete-all", "entity delete"):
        return data

    if command == "entity list":
        result = []
        for r in json_records(data):
            item = pick(r, ["type", "count"])
            item["name"] = r.get("name") or r.get("id", "")
            result.append(item)
        return result

    if command == "event list":
        return [
            pick(r, ["id", "event_type", "status", "latency", "created_at"])
            for r in json_records(data)
        ]

    if command == "event status":
        ev = json_object(data)
        raw_results = json_records(ev.get("results") or [])
        sanitized_results = []
        for r in raw_results:
            nested = r.get("data") or {}
            memory = nested.get("memory") if isinstance(nested, dict) else None
            sanitized_results.append(
                {
                    "id": r.get("id"),
                    "event": r.get("event"),
                    "user_id": r.get("user_id"),
                    "memory": memory,
                }
            )
        event_result = pick(
            ev, ["id", "event_type", "status", "latency", "created_at", "updated_at"]
        )
        event_result["results"] = sanitized_results
        return event_result

    # 直接保留状态、导入和配置命令数据
    return data


def format_agent_envelope(
    console: Console,
    *,
    command: str,
    data: object,
    duration_ms: int | None,
    scope: Mapping[str, JsonValue] | None,
    count: int | None,
) -> None:
    """输出代理模式使用的结构化 JSON 信封。"""
    envelope: JsonObject = {
        "status": "success",
        "command": command,
    }
    if duration_ms is not None:
        envelope["duration_ms"] = duration_ms
    if scope:
        filtered = {k: v for k, v in scope.items() if v}
        if filtered:
            envelope["scope"] = filtered
    if count is not None:
        envelope["count"] = count
    envelope["data"] = json_value(sanitize_agent_data(command, data))

    # 将未认领账号通知放入信封
    # 代理读取 JSON 即可获得提示
    from memgo_cli.runtime.state import take_notice

    notice = take_notice()
    if notice:
        envelope["memgo_notice"] = notice

    console.print_json(json.dumps(envelope, default=str))


def print_result_summary(
    console: Console,
    count: int,
    *,
    duration_secs: float | None,
    page: int | None,
    **scope_ids: str | None,
) -> None:
    """在列表末尾显示结果数量和截断提示。"""
    parts = [f"{count} result{'s' if count != 1 else ''}"]
    if page is not None:
        parts.append(f"page {page}")
    scope_parts = [f"{k}={v}" for k, v in scope_ids.items() if v]
    if scope_parts:
        parts.append(", ".join(scope_parts))
    if duration_secs is not None:
        parts.append(f"{duration_secs:.2f}s")

    summary = " · ".join(parts)
    console.print(f"  [{DIM_COLOR}]{summary}[/]")
    console.print()


def _format_date(dt_str: object) -> str | None:
    """提取日期显示文本。"""
    if not dt_str:
        return None
    try:
        dt = datetime.fromisoformat(json_string(dt_str, "date").replace("Z", "+00:00"))
        return dt.strftime("%Y-%m-%d")
    except (ValueError, AttributeError):
        return str(dt_str)[:10] if dt_str else None
